### ISSUE TITLE
createShortcut but allow any permutation

### ISSUE THREAD

**Issue by onx2** [solidjs-community/solid-primitives#663]
### Describe The Problem To Be Solved

This may not be a bug, and if not I think it would be a nice thing to add as another function or option.

I want to use `createShortcut(["Shift", "Control", "M"], () => {})` and have it trigger the function on any permutation of key presses, so how vscode does it with their shortcuts. 

### Suggest A Solution

I have a rough implementation of it using `useKeyDownList`, so something like this but maybe with also a length check?
```tsx
const keys = useKeyDownList();
const myShortcut = new Set(["SHIFT", "CONTROL", "M"]);

createEffect(() => {
  const keyList = keys();
    for (const key of keyList) {
      if (!myShortcut.has(key)) return;
    }

  someCallback();
});
```

**Comment by thetarnav** [comment]
What is your use case for that?
I’m not entirely opposed to the idea, but it might be a specific need for your app, and like you see is not hard to figure out that functionality yourself.

**Comment by onx2** [comment]
I'm creating a keyboard shortcut manager for an application and want to allow a user to define keys bound to an action. Normally when doing this in software, for example VSCode, Notion, and many others, I can press at least the `Modifier Keys` in any order then a `Character Key` to trigger an action. You can see an example of this in VSCode with `Shift + CMD + P` or in Notion with `Shift + CMD + L`. I actually expected the default behavior of `createShortcut` to do this as it is very commonplace, at least in the software / apps I use. I'm happy to contribute to this as well!

**Comment by shayanhabibi** [comment]
Better that it doesn't do this by default, because it stays true to the principle of a primitive.

If you wanted it to accept any modifier in any order:
Could add an option or a different function which when passed the KBD keys, extracts modifiers into a set. The same is done for events, and then the main functionality of simply comparing the input keyset to the event keyset remains (except now we compare the sequence of character KBDs AND a set of modifier KBDs, instead of a sequence only).

I don't know javascript so I can't contribute this, but I assume it's fairly trivial.

**Comment by onx2** [comment]
> Better that it doesn't do this by default, because it stays true to the principle of a primitive.

I agree to stick with the principle of primitives, however the current implementation prevents extending which is counter-primitive behavior. I actually think it might be best to remove createShortcut if it is going to be opinionated in its implementation (enforcing order). A shortcut utility by nature is opinionated based on design/UX requirements for a specific application, but moreover, enforcing an order is less accessible and extensible than not.

Let me know if the team would like to discuss options or if I should go ahead and close this.

**Comment by atk** [comment]
We could still provide an additional wrapper if we find more applications that have the same use case. I wouldn't want to increase the bundle size of existing apps because of that, though.

**Comment by shayanhabibi** [comment]
I see the point regarding whether the current implementation is overly opinionated.

With regards to this, it should be trivial to remove order on modifiers from what I saw in the source. I prefer against having other keystrokes being unordered because it simplifies the implementation, but more importantly reduces cases where we will be optimistically preventing those keystrokes. That sounds reasonable right?

But then, how would this be reflected in the API? If someone was the input a shortcut code of ctrl S alt X, would it be reasonable to expect order enforcement there?

I get the argument against ordering for modifiers, but would that only apply to any sequence of modifiers in the head of the sequence? Would it apply to any consecutive sequences of modifiers?

**Comment by onx2** [comment]
>But then, how would this be reflected in the API? If someone was the input a shortcut code of ctrl S alt X, would it be reasonable to expect order enforcement there?
I get the argument against ordering for modifiers, but would that only apply to any sequence of modifiers in the head of the sequence? Would it apply to any consecutive sequences of modifiers?

These kinds of questions make me think a shortcut utility doesn't belong as a primitive, but exposing some examples of how it could work is reasonable.

**Addressing Your Questions About Order Enforcement**
My concern isn't just about the order of modifiers, but about any implicit assumptions of order within a `createShortcut` primitive itself.

**Regarding ctrl S alt X and order enforcement**
If a primitive `createShortcut` did exist and allowed for such complex sequences, my view is that enforcing order even there would be overly opinionated for a primitive. The moment we start debating whether alt S ctrl X should trigger ctrl S alt X, we're already entering into application-specific logic/preferences. A primitive should ideally provide the most granular building blocks possible without dictating their specific arrangement or interpretation in complex scenarios. If an application needs to differentiate between ctrl S alt X and alt S ctrl X, that's a valid design choice for that application, but it shouldn't be baked into a low-level primitive.

Regarding the scope of unordered modifiers: If we were to keep `createShortcut` and only relax modifier order, I'd argue it should apply to any set of modifiers concurrently together. This means that for a shortcut like Ctrl + Shift + S, pressing Shift + Ctrl + S should yield the same result.

However, even this "simple" relaxation quickly introduces complexity when we consider sequences. For example, if we have a shortcut A + Ctrl + B, does pressing Ctrl + A + B still trigger it? What about A + B + Ctrl? The distinction between concurrent key presses and sequential key presses becomes crucial, and how a primitive would implicitly handle such variations adds significant opinion.

This further illustrates why these kinds of nuances are better handled by utilities built on top of primitives, rather than as part of the primitive itself.

Zed, for example, allows pressing cmd + s to trigger "save," but if you hold those down and then press shift, it will open the "Save As" functionality. Additionally, holding shift, then pressing s will type an uppercase S, but if you also press cmd while they are down, you still get the "Save As" menu.

I think this behavior might be too nuanced and application-specific for a primitive, especially if we make assumptions about order in any way. I defer to the maintainers though and just wanted to voice those opinions, thanks for humoring me!

**Comment by davedbase** [comment]
@thetarnav or @atk should we close this out?

**Comment by thetarnav** [comment]
I think this is a fine feature request.

**Comment by davedbase** [comment]
Yeah that was simple enough. Thanks for the suggestion @onx2 this will be released in our 2.0 primitives.
